### PR TITLE
new: Spot instances with Managed Node Groups

### PR DESCRIPTION
**Pull request by bijithnair**
#### What this PR does / why we need it:
This PR helps user understands how to setup EKS managed node groups with Spot capacity

#### Which issue(s) this PR fixes:
https://github.com/aws-samples/eks-workshop-v2/issues/455

Fixes #

#### Quality checks

- [x] My content adheres to the style guidelines
- [ ] I ran `make test` or `make e2e-test` and it was successful

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: website/docs/fundamentals/managed-node-groups/spot/index.md [hunk 1]
```diff
@@ -3,21 +3,151 @@ title: Spot instances
...
 
-In addition to the advantages of managed node groups, EKS managed node groups with Spot capacity have these additional advantages:
-
-* Allocation strategy to provision Spot capacity is set to [Capacity Optimized](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-best-practices.html#use-capacity-optimized-allocation-strategy) to ensure that Spot nodes are provisioned in the optimal Spot capacity pools.
-* Specify [multiple instance types](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-best-practices.html#be-instance-type-flexible) during managed node groups creation, to increase the number of Spot capacity pools available for allocating capacity.
+* Allocation strategy to provision Spot capacity is set to Capacity Optimized to ensure that Spot nodes are provisioned in the optimal Spot capacity pools.
```
Comment (svennam92) [inline website/docs/fundamentals/managed-node-groups/spot/index.md:18]: This wording is a bit confusing

File: website/docs/fundamentals/managed-node-groups/spot/index.md [hunk 2]
```diff
@@ -3,21 +3,151 @@ title: Spot instances
...
+
+NAME                                         STATUS   ROLES    AGE     VERSION
+ip-10-42-10-119.us-east-2.compute.internal   Ready    <none>   3d10h   v1.23.15-eks-49d8fe8
+ip-10-42-10-200.us-east-2.compute.internal   Ready    <none>   3d10h   v1.23.15-eks-49d8fe8
+ip-10-42-11-94.us-east-2.compute.internal    Ready    <none>   3d10h   v1.23.15-eks-49d8fe8
+ip-10-42-12-235.us-east-2.compute.internal   Ready    <none>   4h34m   v1.23.15-eks-49d8fe8
+
+:::
+
+In the below diagram, there are two separate Node Group representing the managed node groups within the cluster. The first Node Group box represents the node group containing On-Demand instances while the second represents the node group containing Spot instances, Both are associated with the specified EKS cluster.
```
[the remark this answers is not in the document]
Comment (svennam92) [reply website/docs/fundamentals/managed-node-groups/spot/index.md:49]: > Spot instances, Both

, -> .

File: website/docs/fundamentals/managed-node-groups/spot/index.md [hunk 3]
```diff
@@ -3,21 +3,151 @@ title: Spot instances
...
+
+![spot arch](../assets/managed-spot-arch.png)
+
+As our existing cluster already has a nodegroup with `On-Demand` instances, The next step would be to setup a node group which has EC2 instances with capacity type as `SPOT`. 
+
+To achieve that, We will do following:
+
+* Export the Managed Node Group name for SPOT
+
+* Use AWS CLI to create EKS managed node group for SPOT
```
Comment (svennam92) [inline website/docs/fundamentals/managed-node-groups/spot/index.md:59]: Spot or SPOT

File: website/docs/fundamentals/managed-node-groups/spot/index.md [hunk 4]
```diff
@@ -3,21 +3,151 @@ title: Spot instances
...
+
+:::
+
+In the below diagram, there are two separate Node Group representing the managed node groups within the cluster. The first Node Group box represents the node group containing On-Demand instances while the second represents the node group containing Spot instances, Both are associated with the specified EKS cluster.
+
+![spot arch](../assets/managed-spot-arch.png)
+
+As our existing cluster already has a nodegroup with `On-Demand` instances, The next step would be to setup a node group which has EC2 instances with capacity type as `SPOT`. 
+
+To achieve that, We will do following:
```
[the remark this answers is not in the document]
Comment (svennam92) [reply website/docs/fundamentals/managed-node-groups/spot/index.md:55]: do the following two lines need to be bullet points?

File: website/docs/fundamentals/managed-node-groups/spot/index.md [hunk 5]
```diff
@@ -3,21 +3,151 @@ title: Spot instances
...
+--node-role $MANAGED_NODE_GROUP_IAM_ROLE_ARN \
+--subnets $PRIMARY_SUBNET_1 $PRIMARY_SUBNET_2 $PRIMARY_SUBNET_3 \
+--instance-types m5.large m5d.large m5a.large m5ad.large m5n.large m5dn.large \
+--capacity-type SPOT \
+--scaling-config minSize=2,maxSize=6,desiredSize=2 \
+--disk-size 20 \
+--labels capacity_type=managed_spot
+
+```
+To track the status of Node Group creation, You can running below command in a separate terminal.
```
Comment (svennam92) [inline website/docs/fundamentals/managed-node-groups/spot/index.md:76]: grammar

File: website/docs/fundamentals/managed-node-groups/spot/index.md [hunk 6]
```diff
@@ -3,21 +3,151 @@ title: Spot instances
...
+ip-10-42-10-96.us-west-2.compute.internal    Ready    <none>   113m   v1.23.15-eks-49d8fe8   ON_DEMAND      managed-ondemand-20230605211738568600000028
+ip-10-42-11-17.us-west-2.compute.internal    Ready    <none>   78s    v1.23.17-eks-0a21954   SPOT           managed-spot
+ip-10-42-12-234.us-west-2.compute.internal   Ready    <none>   77s    v1.23.17-eks-0a21954   SPOT           managed-spot
+ip-10-42-12-45.us-west-2.compute.internal    Ready    <none>   113m   v1.23.15-eks-49d8fe8   ON_DEMAND      managed-ondemand-20230605211738568600000028
+
+```
+From the above output you can see that now we have two managed nodegroups available. Next, let’s deploy the “Sample Retail Store” app and leverage [nodeSelector](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector) as node selection constraint to deploy app on top of the spot instances instead of On-Demand. Using the nodeSelector field, you can specify key-value pairs to filter and select which nodes in the cluster should run particular pods. It is used to define constraints for pod placement based on node labels.
```
Comment (svennam92) [inline website/docs/fundamentals/managed-node-groups/spot/index.md:113]: wrap `nodeSelector` with `

File: website/docs/fundamentals/managed-node-groups/spot/index.md [hunk 7]
```diff
@@ -3,21 +3,151 @@ title: Spot instances
...
+ip-10-42-12-234.us-west-2.compute.internal   Ready    <none>   77s    v1.23.17-eks-0a21954   SPOT           managed-spot
+ip-10-42-12-45.us-west-2.compute.internal    Ready    <none>   113m   v1.23.15-eks-49d8fe8   ON_DEMAND      managed-ondemand-20230605211738568600000028
+
+```
+From the above output you can see that now we have two managed nodegroups available. Next, let’s deploy the “Sample Retail Store” app and leverage [nodeSelector](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector) as node selection constraint to deploy app on top of the spot instances instead of On-Demand. Using the nodeSelector field, you can specify key-value pairs to filter and select which nodes in the cluster should run particular pods. It is used to define constraints for pod placement based on node labels. 
+
+Now because our existing manifest `deployment.yaml` located under `/workspace/manifests/catalog`  does not have `nodeSelector` as an attribute, we will use [kustomize](https://www.eksworkshop.com/docs/introduction/kustomize/) to modify the existing resource configuration without directly modifying the original manifests.
```
Comment (svennam92) [inline website/docs/fundamentals/managed-node-groups/spot/index.md:115]: A bit wordy. Can simply say "Run the following command to update the existing deployment.yaml to add `nodeSelector`.

File: website/docs/fundamentals/managed-node-groups/spot/index.md [hunk 8]
```diff
@@ -3,21 +3,151 @@ title: Spot instances
...
+From the above output you can see that now we have two managed nodegroups available. Next, let’s deploy the “Sample Retail Store” app and leverage [nodeSelector](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector) as node selection constraint to deploy app on top of the spot instances instead of On-Demand. Using the nodeSelector field, you can specify key-value pairs to filter and select which nodes in the cluster should run particular pods. It is used to define constraints for pod placement based on node labels. 
+
+Now because our existing manifest `deployment.yaml` located under `/workspace/manifests/catalog`  does not have `nodeSelector` as an attribute, we will use [kustomize](https://www.eksworkshop.com/docs/introduction/kustomize/) to modify the existing resource configuration without directly modifying the original manifests. 
+
+```kustomization
+fundamentals/mng/spot/deployment.yaml
+Deployment/catalog
+```
+
+Next, Deploy the app.
```
Comment (svennam92) [inline website/docs/fundamentals/managed-node-groups/spot/index.md:122]: Grammar

File: website/docs/fundamentals/managed-node-groups/spot/index.md [hunk 9]
```diff
@@ -3,21 +3,151 @@ title: Spot instances
...
+serviceaccount/catalog created
+configmap/catalog created
+secret/catalog-db created
+service/catalog created
+service/catalog-mysql created
+deployment.apps/catalog created
+statefulset.apps/catalog-mysql created
+```
+
+To make sure your app got deployed successfully, We will use `kubectl rollout status` command that allows you to check the status of a deployment in Kubernetes. It provides information about the progress of the rollout and the current state of the associated resources.
```
Comment (svennam92) [inline website/docs/fundamentals/managed-node-groups/spot/index.md:137]: Grammar

File: website/docs/fundamentals/managed-node-groups/spot/index.md [hunk 10]
```diff
@@ -3,21 +3,151 @@ title: Spot instances
...
+```bash
+$ kubectl rollout status deployment/catalog -n catalog --timeout=5m
+```
+Once the application get’s deployed successfully, The Final step would be to check whether the app got deployed on `SPOT` instances, To do that run below script on your terminal.
```
Comment (svennam92) [inline website/docs/fundamentals/managed-node-groups/spot/index.md:141]: Grammar

File: website/docs/fundamentals/managed-node-groups/spot/index.md [hunk 11]
```diff
@@ -3,21 +3,151 @@ title: Spot instances
...
+
+```bash
+$ for pod in $(kubectl get pods -l app.kubernetes.io/name=catalog -n $CATALOG_RDS_DATABASE_NAME  | awk 'NR>1 {print $1}'); do if [ "$pod" != "NAME" ]; then node=$(kubectl get pod $pod -o json -n $CATALOG_RDS_DATABASE_NAME | jq -r '.spec.nodeName'); node_group=$(kubectl get node $node -o json | jq -r '.metadata.labels."eks.amazonaws.com/capacityType"'); echo "$pod | $node | $node_group"; fi; done
```
Comment (svennam92) [inline website/docs/fundamentals/managed-node-groups/spot/index.md:144]: Instead of a custom command, can we do a combination of `$ kubectl get nodes -l eks.amazonaws.com/capacityType=ON_DEMAND` like above, followed by a `kubectl get pods`:

`kubectl get pod -o=custom-columns=NAME:.metadata.name,STATUS:.status.phase,NODE:.spec.nodeName --all-namespaces`

User can connect the dots.

**Review by svennam92 — CHANGES_REQUESTED** [review]
Solid initial PR, some slight changes before we can merge

**Comment by bijithnair** [comment]
Due to the recent changes to the workshop, Kustomize location got changed and need fix

**Review by svennam92 — DISMISSED** [review]

File: manifests/modules/fundamentals/mng/spot/deployment.yaml [hunk 12]
```diff
@@ -0,0 +1,9 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: catalog
+spec:
+  template:
+    spec:
+      nodeSelector:
+        capacity_type: managed_spot
```
Comment (niallthomson) [inline manifests/modules/fundamentals/mng/spot/deployment.yaml:9]: Can this use the built-in lab instead of a custom one? Shouldn't be needed.

**Review by niallthomson — COMMENTED** [review]
A couple of general questions before I really dig in to the content.

First, I see a `prepare-environment` on the index page, but no lab indicator. I'm thinking that perhaps we should just re-structure the MNG lab in to several separate labs. The existing content goes in something like "MNG basics", then we add standalone labs for Spot and Graviton2. I think its likely that customers might want to skip the easy stuff but still learn about the latter.

Second, it would be good to cover instance diversification since this is pretty critical with Spot in MNG. We have this outlined in the Finhack workshop: https://catalog.us-east-1.prod.workshops.aws/workshops/c4ab40ed-0299-4a4e-8987-35d90ba5085e/en-US/050-ec2-spot-instances/selecting-instance-types

**Comment by niallthomson** [comment]
@bijithnair thanks for all the work you put in to this content!
